File: src/main/java/com/github/zxh/classpy/classfile/constant/ConstantFactory.java
package com.github.zxh.classpy.classfile.constant;

import com.github.zxh.classpy.common.ParseException;

public class ConstantFactory {

    /**
     * Create concrete ConstantXxxInfo by tag.
     * @param tag
     * @return
     */
    public static ConstantInfo create(byte tag) {
        switch (tag) {
            case  1: return new ConstantUtf8Info();
            case  3: return new ConstantIntegerInfo();
            case  4: return new ConstantFloatInfo();
            case  5: return new ConstantLongInfo();
            case  6: return new ConstantDoubleInfo();
            case  7: return new ConstantClassInfo();
            case  8: return new ConstantStringInfo();
            case  9: return new ConstantFieldrefInfo();
            case 10: return new ConstantMethodrefInfo();
            case 11: return new ConstantInterfaceMethodrefInfo();
            case 12: return new ConstantNameAndTypeInfo();
            case 15: return new ConstantMethodHandleInfo();
            case 16: return new ConstantMethodTypeInfo();
            case 18: return new ConstantInvokeDynamicInfo();
            case 19: return new ConstantModuleInfo();
            case 20: return new ConstantPackageInfo();
        }

        throw new ParseException("Invalid Constant Type: " + tag);
    }

}

File: src/main/java/com/github/zxh/classpy/classfile/constant/ConstantFactory.kt
package com.github.zxh.classpy.classfile.constant;

import com.github.zxh.classpy.classfile.ClassParseException;

object  ConstantFactory {

    /**
     * Create concrete ConstantXxxInfo by tag.
     * @param tag
     * @return
     */
    fun create(tag: Byte): ConstantInfo {
        when (tag.toInt()) {
             1 -> return ConstantUtf8Info();
             3 -> return ConstantIntegerInfo();
             4 -> return ConstantFloatInfo();
             5 -> return ConstantLongInfo();
             6 -> return ConstantDoubleInfo();
             7 -> return ConstantClassInfo();
             8 -> return ConstantStringInfo();
             9 -> return ConstantFieldrefInfo();
            10 -> return ConstantMethodrefInfo();
            11 -> return ConstantInterfaceMethodrefInfo();
            12 -> return ConstantNameAndTypeInfo();
            15 -> return ConstantMethodHandleInfo();
            16 -> return ConstantMethodTypeInfo();
            18 -> return ConstantInvokeDynamicInfo();
        }

        throw ClassParseException("Invalid Constant Type: " + tag);
    }

}
